Fix race condition in cinder attach/detach

jsafrane opened the pull request:
Resolves https://github.com/kubernetes/kubernetes/issues/19602

Most of the changes are related to unit test. The real fix is in cinder.go. Similar change is needed in most of the volume plugins! Cinder seems to be the easiest to reproduce in real world, as it's the slowest to attach volumes (and thus the window of opportunity is the largest).

Also, I added some logs here and there to debug this issue easily next time.

jsafrane commented:
@k8s-bot unit test this

jsafrane commented:
@k8s-bot test this

jsafrane commented:
@k8s-bot test this

markturansky commented on pkg/volume/cinder/cinder_test.go:
```diff
@@ -55,20 +60,62 @@ func TestCanSupport(t *testing.T) {
+	}
+
+	// Cinder detach is slow...
+	//time.Sleep(fake.attachDetachDuration)
```
left over sleep statement.  I assume this helped you find the race?

jsafrane replied:
Indeed, it's a left over from debugging. It's not needed to reproduce the bug, I removed it.

saad-ali commented on pkg/volume/util.go:
```diff
@@ -148,3 +149,66 @@ func CalculateTimeoutForVolume(minimumTimeout, timeoutIncrement int, pv *api.Per
+// SetUp and TearDown operation on the same volume. Most plugins are not
+// thread safe and bad things can happend, when they SetUp and TearDown
+// the same volume at the same time. With a VolumeGuard, they are thread safe.
+type VolumeGuard struct {
```
This is very similar to [`pkg/util/keymutex/keymutex.go` ](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gce_pd/gce_util.go) which is used for the same purpose for GCE PDs. It would make sense to reuse that code and modify it if needed.

jsafrane replied:
Thanks for finding this, I removed my `VolumeGuard` completely.

jsafrane commented:
Rebased & removed custom `VolumeGuard` mutexes.

saad-ali commented:
LGTM. Thanks for the fix @jsafrane. Please squash the commits and I'll add the LGTM label.

jsafrane commented on pkg/volume/cinder/cinder_test.go:
```diff
@@ -191,9 +236,123 @@ func TestPlugin(t *testing.T) {
+// Test a race when a volume is simultaneously SetUp and TearedDown
+func TestAttachDetachRace(t *testing.T) {
+	flag.Parse()
+	flag.Lookup("v").Value.Set("10")
```
Oops, some debugging leftovers, removed from the final patch.

jsafrane commented:
@saad-ali, I *can* squash them, but I actually prefer the way they are organized now. Small patches with clear purpose.

markturansky commented:
@jsafrane clear to review, yes, and that is helpful.  But still a logical single commit that is easy to rollback if things go south.  Squashing before merging a PR is standard procedure.

jsafrane commented:
Squashed

markturansky commented:
Thanks, Jan.  I'll add LGTM, since @saad-ali said he would have done that after a squash.

markturansky commented on pkg/volume/cinder/cinder.go:
```diff
@@ -35,11 +36,13 @@ import (
 type cinderPlugin struct {
 	host volume.VolumeHost
+	// Guarding SetUp and TearDown operations
+	volumeGuard keymutex.KeyMutex
```
Rename to "lock" or "mutex"?  I've seen those two names throughout the codebase. I've used both in the pv controllers, but should probably standardize on one.

I prefer "mutex" (since that's what the thing actually is) and I will be sure to change "lock" to "mutex" in the controllers as we refactor them.

markturansky replied:
Also, I'm looking at KeyMutex and this is the exact thing we need to stripe locks by pv/pvc name.

jsafrane replied:
renamed to `volumeLocks`.

markturansky replied:
i don't see the change.  also, how about just `locks`?   `b.plugin.locks.LockKey(b.pdName)` ?

jsafrane replied:
Sorry, pushed now.

jsafrane commented:
```
=== RUN   TestAttachDetachRace
I0121 15:37:58.424363   19808 cinder_test.go:301] Attaching volume
I0121 15:37:58.924502   19808 cinder_test.go:312] Detaching volume
==================
WARNING: DATA RACE
Write by goroutine 19:
  k8s.io/kubernetes/pkg/util/mount.(*FakeMounter).Mount()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/mount/fake.go:70 +0x3cc
  k8s.io/kubernetes/pkg/volume/cinder.(*cinderVolumeBuilder).SetUpAt()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder.go:271 +0x10dd
  k8s.io/kubernetes/pkg/volume/cinder.(*cinderVolumeBuilder).SetUp()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder.go:231 +0x7b
  k8s.io/kubernetes/pkg/volume/cinder.TestAttachDetachRace.func1()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder_test.go:302 +0x74

Previous read by goroutine 18:
  k8s.io/kubernetes/pkg/util/mount.(*FakeMounter).IsLikelyNotMountPoint()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/mount/fake.go:95 +0x64
  k8s.io/kubernetes/pkg/volume/cinder.(*cinderVolumeCleaner).TearDownAt()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder.go:328 +0x210
  k8s.io/kubernetes/pkg/volume/cinder.(*cinderVolumeCleaner).TearDown()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder.go:321 +0x7b
  k8s.io/kubernetes/pkg/volume/cinder.TestAttachDetachRace()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder_test.go:313 +0x1479
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc
```

I'll add locks around but come on, it's fake mounter...

jsafrane commented:
@k8s-bot unit test this.

I hate jenkins:
> Unable to perform initial IP allocation check: unable to persist the updated service IP allocations: serviceipallocations "" cannot be updated: the provided resource version does not match

markturansky commented on pkg/util/mount/fake.go:
```diff
@@ -37,20 +46,56 @@ type FakeAction struct {
 		if mp.Path != target {
 			newMountpoints = append(newMountpoints, MountPoint{Device: mp.Device, Path: mp.Path, Type: mp.Type})
+		} else {
+			glog.V(5).Infof("Fake mounter: unmouted %s from %s", mp.Device, target)
```
This if/else is clear to me, but I'm wondering about the Go (and Kube) preference to un-indent control flow.  Is something like this better or worse for idiomatic Go?

```
for {
    if mp.Path != target {
        newMountpoints = append(newMountpoints, MountPoint{...})
        continue 
    }
    glog.V(5).Infof("logstatement")
}
```

I'm not saying to make this change.  This is an open question as to style and idiomatic Go that I'd like to understand better.

@pmorie @smarterclayton any pointers on stylistic preference?

pmorie replied:
I have no personal preference on this matter of style and I'm sure both styles exist in the codebase.

smarterclayton replied:
We prefer not to use elses when continues accomplish the same function.  We also prefer the "meat" of the loop to be at the end of the loop block and the checks before to invoke continue for exceptional cases:

```
for {
  ... find reasons to exit early
  if exceptionalCase {
    continue
  }
  ... whatever the point of the loop is
  b = append(b, ...)
}
```

markturansky replied:
Thanks @smarterclayton.  That was my understanding, as well.  I've been asked in different PRs to drop `else` from many `if` statements and the flow still made sense.

markturansky commented on pkg/volume/cinder/cinder.go:
```diff
@@ -309,37 +324,65 @@ func (c *cinderVolumeCleaner) TearDown() error {
+
+	// Find Cinder volumeID to lock the right volume
+	// TODO: refactor VolumePlugin.NewCleaner to get full volume.Spec just like
+	// NewBuilder. We could then find volumeID there without probing MountRefs.
```
Tim has this same kind of TODO in kubelet's volume code:  https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumes.go#L216

Not for this PR, of course, but it's worth thinking about a solution.

jsafrane replied:
I think kubelet has whole VolumeSpec available when it creates cleaners, so it's fairly simple change. On the other hand, all the volume plugins needs to be updated with new `NewCleaner` signature.

markturansky commented on pkg/volume/cinder/cinder.go:
```diff
@@ -309,37 +324,65 @@ func (c *cinderVolumeCleaner) TearDown() error {
+	c.plugin.volumeLocks.LockKey(c.pdName)
+	defer c.plugin.volumeLocks.UnlockKey(c.pdName)
+
+	// Reload list of references, there might be SetUpAt finished in the meantime
```
if the lock is at the beginning of the function, would you still need to reload the list here?

jsafrane replied:
All the code before `LockKey(c.pdName)` is there to actually find `c.pdName`. That's why passing whole `VolumeSpec` to `NewCleaner()` would be **much** nicer - `c.pdName` is actually `VolumeSpec.Cinder.VolumeID`.

jsafrane commented:
Rebased

markturansky commented:
LGTM.

Since @saad-ali also LGTM'd it, I will add the label.

eparis commented:
@k8s-bot test this issue: #IGNORE

Tests have been pending for 24 hours

jsafrane committed:
Fixed races in Cinder volume attach/detach.

Add a mutex to guard SetUpAt() and TearDownAt() calls - they should not
run in parallel.  There is a race in these calls when there are two pods
using the same volume, one of them is dying and the other one starting.

TearDownAt() checks that a volume is not needed by any pods and detaches the
volume. It does so by counting how many times is the volume mounted
(GetMountRefs() call below).

When SetUpAt() of the starting pod already attached the volume and did not mount
it yet, TearDownAt() of the dying pod will detach it - GetMountRefs() does not
count with this volume.

These two threads run in parallel:

 dying pod.TearDownAt("myVolume")          starting pod.SetUpAt("myVolume")
   |                                       |
   |                                       AttachDisk("myVolume")
   refs, err := mount.GetMountRefs()       |
   Unmount("myDir")                        |
   if refs == 1 {                          |
   |  |                                    Mount("myVolume", "myDir")
   |  |                                    |
   |  DetachDisk("myVolume")               |
   |                                       start containers - OOPS! The volume is detached!
   |
   finish the pod cleanup


Also, add some logs to cinder plugin for easier debugging in the future, add
a test and update the fake mounter to know about bind mounts.

jsafrane commented:
@markturansky, I had to rebase and I lost LGTH label, can you please add it again?
